### ISSUE TITLE
Rule await-promise: make Promise implementation configurable

### ISSUE THREAD

**Issue by darxriggs** [palantir/tslint#2661]
### Feature Request

- __TSLint version__: 5.1.0
- __TypeScript version__: 2.3.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
import * as Bluebird from 'bluebird';

function functionReturningThenable(): Bluebird.Thenable<string> {
    return Bluebird.resolve('some data');
}

async function main(): Promise<void> {
    const data = await functionReturningThenable();
    console.log(data);
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "await-promise": true
  }
}
```

#### Actual behavior
```shell
(await-promise) src/test.ts[8, 18]: 'await' of non-Promise.
```

#### Expected behavior
No error should be reported.

There are multiple Promise implementations and interfaces out there (`JQueryPromise`, `Q.Promise`, `Thenable`, etc.). Also see [Promises/A+](https://promisesaplus.com/), Therefore the rule should be configurable like `no-floating-promises` already is.

**Comment by darxriggs** [comment]
I'd like to create PR for this but I've seen https://github.com/palantir/tslint/pull/2631 ("no-floating-promises: rewrite as walker function").

So the question is if this rule should use `applyWithWalker` or `applyWithFunction` and why?

**Comment by ajafff** [comment]
@darxriggs we want to move away from `RuleWalker` and therefore `ProgramAwareRuleWalker` and eventually deprecate them. `await-promise` has not been refactored to use either `AbstractWalker` or `applyWithFunction`. #2631 would be a good start.

Which one you choose is up to you. For simple rules it's preferred to use a function. When you find yourself in the need to add several closures to the walker function, you should consider using `AbstractWalker` to avoid the repeated closure allocation.
For more information about the concepts, there is some documentation at https://palantir.github.io/tslint/develop/custom-rules/walker-design.html
